File: python/mirage/commands/builtin/generic_bind/builders/truncate.py
from functools import partial

from mirage.accessor.base import Accessor
from mirage.cache.index import NULL_INDEX, IndexCacheStore
from mirage.commands.builtin.generic.truncate import \
    truncate as generic_truncate
from mirage.commands.builtin.generic_bind.adapter import (Builder, CommandIO,
                                                          Operation, bound_op)
from mirage.commands.spec import SPECS
from mirage.commands.spec.types import FlagView
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec


async def truncate(
    ops: CommandIO,
    accessor: Accessor,
    paths: list[PathSpec],
    *texts: str,
    stdin: ByteSource | None = None,
    s: str | None = None,
    index: IndexCacheStore = NULL_INDEX,
    **flags: object,
) -> tuple[ByteSource | None, IOResult]:
    fl = FlagView(flags, spec=SPECS["truncate"])
    size = s or fl.as_str("size")
    if size is None:
        raise ValueError("truncate: you must specify either '--size' or '-s'")
    truncate_fn = ops.require(Operation.TRUNCATE)
    paths = await ops.resolve_glob(accessor, paths, index)
    return await generic_truncate(
        paths,
        size=size,
        stat=bound_op(ops.stat, accessor, index),
        truncate_fn=partial(truncate_fn, accessor),
    )


BUILDER = Builder("truncate",
                  truncate,
                  write=True,
                  requirements=frozenset({Operation.TRUNCATE}))

File: typescript/packages/core/src/commands/builtin/generic_bind/builders/truncate.ts
import { truncateGeneric } from '../../generic/truncate.ts'
import { specOf } from '../../../spec/builtins.ts'
import { FlagView } from '../../../spec/types.ts'
import { type Builder, resolveGlobOf } from '../adapter.ts'

export const TRUNCATE_BUILDER: Builder = {
  name: 'truncate',
  write: true,
  requirements: ['truncate'],
  fn: async (ops, accessor, paths, _texts, opts) => {
    const sizeValue = new FlagView(opts.flags, specOf('truncate')).asStr('size')
    if (sizeValue === undefined) {
      throw new Error("truncate: you must specify either '--size' or '-s'")
    }
    const { truncate } = ops
    if (truncate === undefined) throw new Error('truncate: backend provides no truncate op')
    const index = opts.index ?? undefined
    const resolved = await resolveGlobOf(ops)(accessor, paths, index)
    return truncateGeneric(
      resolved,
      sizeValue,
      (path) => ops.stat(accessor, path, index),
      (path, length) => truncate(accessor, path, length),
    )
  },
}
